docs/config.rst

```diff
@@ -0,0 +1,180 @@
+    node-directory. It would probably still be convenient to refer to
+    a Tahoe client by node-directory (but we can just pull out the
+    Web-API endpoint). If we only store the endpoint and the Tahoe
+    config changes, the user would also have to change their
+    magic-folder config.
+- our own API endpoint (a Twisted server endpoint-string,
+  e.g. `tcp:1234` or `unix:/tmp/foo`)
+
+State that we need to store:
+
+- a secret token that protects the API (only users who can read the
+  file can access the endpoint)
+- list of other participants (per magic-folder) (see note)
+- local cache of remote snapshots (probably per-magic folder)
+- local copy of snapshots we haven't yet uploaded (probably per-magic folder)
+  - metadata details of the snapshot
+  - "stash directory" for file-data of the snapshot
+
+NOTE: Currently the "list of other participants" comes from the
+"collective dircap" but in the Leif Design each user decides whom they
+wish to pull updates from. So, the "list of other participants" above
+might be "a readcap of a directory" (and we would determine the actual
+list at runtime by reading that capability).
+
+
+UI to Change the Config
+-----------------------
+
+Writing configuration files is hard, error-prone and not a great
+interface. It also makes it hard (or impossible) to change where or
+how configuration is stored. All configuration shall be changed by an
+HTTP API. There will also be CLI commands that are thin wrappers of
+the HTTP API. GUI front-ends could use the command-line or the HTTP
+API. Nothing except the HTTP API should be touching the actual stored
+configuration.
+
+(Of course we can't stop users from editing the config files
+themselves, but this should be discouraged).
+
+
+.. _current:
+
+Where and How To Store the Configuration
+----------------------------------------
+
+Somewhat mirroring Tahoe-LAFS's design, configuration shall all be
+found in a single directory. This directory can be specified via
+command-line option. (Should we have a default? Seems sensible, most
+OSes have a reasonably well-defined place for programs to store
+configuration data).
+
+All global configuration and state shall be stored in an SQLite3
+database. This database will have tables for global configuration and
+state.
+
+The "configuration directory" will look like this:
+
+- `confdir` (e.g. `~/.config/magic-folder`)
+  - `global.sqlite`
+  - `api_token`: 32 bytes of binary data
+  - `magic_folder.pid`: contains the PID of the currently-running process
+
+The `global.sqlite` database will have the following tables:
+
+- "version" with a single row containing "1"
+- "magic_folders"
+  - columns: `name`, `location`
+  - `location` is a local path for that folder's configuration
+- "config"
+  - columns: "api_endpoint", "tahoe_client_url"
+  - `api_endpoint` is a Twisted server string description
+  - `tahoe_client_url` is a URL to our Tahoe-LAFS client's Web-UI
```

Comment:
I guess the reason we might want the node directory instead of the http api root url is that some nodes are configured to listen on a random port  every startup?  And, specifically, gridsync configures its tahoe nodes to do this.  It might be annoying to have to update database state every time the data node restarts.

Or, we could mitigate this by *allowing* `tahoe_client_url` as a cli argument to `magic-folder run` and having it override the database value?

Or we could have a node directory in the db and read the node.url ourselves...

Or some combination of the above, I guess.

Reply:
Pointing at the node directory means the magic-folder process could re-read a changed node.url itself and follow tahoe-lafs restarts/reconfigurations easily.

Reply:
okay, sounds good. also storing node-directory is definitely *safer* too (e.g. for anything else we forgot about that we need from tahoe still).